### ISSUE TITLE
java.nio.file.ClosedFileSystemException

### ISSUE THREAD

**Issue by reid-spencer** [scala-native/scala-native#4101]
## Project 
My project is a multi-module, cross-project set of libraries and a cli that builds to JVM, JS, and Native. 

## Symptom
When I do a full (parallel) build on linux or MacOS, I get three of the six native modules failing with a `java.nio.file.ClosedFileSystemException` exception while trying to build the test executable.   But,  when just `sbt test` is run on any one native module, to narrow the reproduction scenario, the exception does not happen.  

## Context
* ScalaNative: 0.5.6
* JVM: OpenJDK 21.0.2
* SBT: 1.10.6
* Platforms: 
  * MacOS 15.1.1 running Clang 19.1.5 for arm64-apple-darwin24.1.0
  * Ubuntu 24.04 running Clang  14.0.0-1ubuntu1.1 for x86_64-pc-linux-gnu
* The JVM has 8GB of memory in which to build this project (4GB has usually been enough before adding scala-native support to the project)
* The native compilation configuration is governed by [this sbt configuration function](https://github.com/ossuminc/sbt-ossuminc/blob/c8db5a89b7db1507de29efe7061a63235869c456/src/main/scala/com/ossuminc/sbt/helpers/Native.scala#L24) with all default arguments. Feel free to check the implementation, but the equivalent `nativeConfig` is:

```
Compile / nativeConfig ~= { c => c
  .withLTO(LTO.none)
  .withMode(Mode.releaseFast)
  .withGC(GC("immix"))
  .withBuildTarget(BuildTarget.libraryStatic)
  .withTargetTriple(None)
  .withCompileOptions(c.compileOptions ++ Seq.empty)
  .withLinkingOptions(c.linkingOptions ++ Seq.empty)
  .withEmbedResources(true)
  .withOptimize(true)
}
``` 

## Reproduction Steps
* git clone https://github.com/ossuminc/riddl.git
* git checkout native-build
* sbt -v clean Test/compile test "project riddlc" Universal/packageBin \
          "project riddlcNative" nativeLink

It is likely to fail during the "test" operation, so the last command line could be shorter. 

## Potential Workarounds
* Turn parallel testing off in sbt (will dramatically raise the testing time)
* Provide even more memory
* Run the native test projects sequentially instead of in parallel

## Analysis
The OpenJDK documentation about the  raised exception says:
> Unchecked exception thrown when an attempt is made to invoke an operation on a file and the file system is closed.

It isn't clear at all WHAT "file system" is closed. An operating system file system,  or the "ZipFileSystem" involved that produces the exception?  It is likely, however, to be the latter. I am wondering if the ScalaNative "VirtualDirectory" shown in the call stack is perhaps a factor in the file system being closed?

From issue #3595, we know that unchecked exceptions at runtime may be problematic, but this issue does not have that symptom. 

## Possible resolutions: 
* Check for a closed "file system" before trying to use it.
* Replace the call with `Files.readAllBytes` with a streaming solution, perhaps?
* Handle the exception more gracefully and report the actual issue instead of allowing the exception. 

## Exception Backtrace
```
[error] java.nio.file.ClosedFileSystemException
[error]         at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.ensureOpen(ZipFileSystem.java:1635)
[error]         at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.newByteChannel(ZipFileSystem.java:971)
[error]         at jdk.zipfs/jdk.nio.zipfs.ZipPath.newByteChannel(ZipPath.java:870)
[error]         at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.newByteChannel(ZipFileSystemProvider.java:247)
[error]         at java.base/java.nio.file.Files.newByteChannel(Files.java:379)
[error]         at java.base/java.nio.file.Files.newByteChannel(Files.java:431)
[error]         at java.base/java.nio.file.Files.readAllBytes(Files.java:3268)
[error]         at scala.scalanative.io.VirtualDirectory$NioDirectory.read(VirtualDirectory.scala:95)
[error]         at scala.scalanative.io.VirtualDirectory$NioDirectory.read$(VirtualDirectory.scala:94)
[error]         at scala.scalanative.io.VirtualDirectory$JarDirectory.read(VirtualDirectory.scala:172)
[error]         at scala.scalanative.nir.serialization.package$.deserializeBinary(package.scala:24)
[error]         at scala.scalanative.linker.ClassPath$Impl.$anonfun$load$2(ClassPath.scala:79)
[error]         at scala.Option.map(Option.scala:230)
[error]         at scala.scalanative.linker.ClassPath$Impl.$anonfun$load$1(ClassPath.scala:78)
[error]         at scala.collection.mutable.HashMap.getOrElseUpdate(HashMap.scala:86)
[error]         at scala.scalanative.linker.ClassPath$Impl.load(ClassPath.scala:78)
[error]         at scala.scalanative.linker.ClassLoader$FromDisk$$anonfun$load$1.applyOrElse(ClassLoader.scala:42)
[error]         at scala.scalanative.linker.ClassLoader$FromDisk$$anonfun$load$1.applyOrElse(ClassLoader.scala:40)
[error]         at scala.collection.TraversableOnce.collectFirst(TraversableOnce.scala:180)
[error]         at scala.collection.TraversableOnce.collectFirst$(TraversableOnce.scala:167)
[error]         at scala.collection.AbstractTraversable.collectFirst(Traversable.scala:108)
[error]         at scala.scalanative.linker.ClassLoader$FromDisk.load(ClassLoader.scala:40)
[error]         at scala.scalanative.linker.Reach.lookup(Reach.scala:135)
[error]         at scala.scalanative.linker.Reach.lookup(Reach.scala:126)
[error]         at scala.scalanative.linker.Reach.reachDefn(Reach.scala:196)
[error]         at scala.scalanative.linker.Reach.reachGlobalNow(Reach.scala:320)
[error]         at scala.scalanative.linker.Reach.classInfo(Reach.scala:505)
[error]         at scala.scalanative.linker.Reach.reachOp(Reach.scala:790)
[error]         at scala.scalanative.linker.Reach.reachInst(Reach.scala:723)
[error]         at scala.scalanative.linker.Reach.$anonfun$reachInsts$1(Reach.scala:715)
[error]         at scala.scalanative.linker.Reach.$anonfun$reachInsts$1$adapted(Reach.scala:715)
[error]         at scala.collection.Iterator.foreach(Iterator.scala:943)
[error]         at scala.collection.Iterator.foreach$(Iterator.scala:943)
[error]         at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
[error]         at scala.collection.IterableLike.foreach(IterableLike.scala:74)
[error]         at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
[error]         at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
[error]         at scala.scalanative.linker.Reach.reachInsts(Reach.scala:715)
[error]         at scala.scalanative.linker.Reach.reachDefine(Reach.scala:628)
[error]         at scala.scalanative.linker.Reach.reachDefn(Reach.scala:223)
[error]         at scala.scalanative.linker.Reach.$anonfun$reachDefn$2(Reach.scala:202)
[error]         at scala.scalanative.linker.Reach.$anonfun$reachDefn$2$adapted(Reach.scala:198)
[error]         at scala.Option.fold(Option.scala:251)
[error]         at scala.scalanative.linker.Reach.reachDefn(Reach.scala:198)
[error]         at scala.scalanative.linker.Reach.process(Reach.scala:159)
[error]         at scala.scalanative.linker.Reach$.apply(Reach.scala:1167)
[error]         at scala.scalanative.linker.Link$.apply(Link.scala:12)
[error]         at scala.scalanative.build.ScalaNative$.$anonfun$link$3(ScalaNative.scala:45)
[error]         at scala.scalanative.build.Logger.time(Logger.scala:36)
[error]         at scala.scalanative.build.Logger.time$(Logger.scala:33)
[error]         at scala.scalanative.build.Logger$$anon$3.time(Logger.scala:102)
[error]         at scala.scalanative.build.ScalaNative$.$anonfun$link$1(ScalaNative.scala:45)
[error]         at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
[error]         at scala.util.Success.$anonfun$map$1(Try.scala:255)
[error]         at scala.util.Success.map(Try.scala:213)
[error]         at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
[error]         at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:42)
[error]         at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:74)
[error]         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
[error]         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
[error]         at java.base/java.lang.Thread.run(Thread.java:1583)
```

**Comment by JD557** [comment]
I believe this is a duplicate of https://github.com/scala-native/scala-native/issues/2024

**Comment by reid-spencer** [comment]
@JD557 - Is your suggestion an sbt expression placed in the settings for the project using the `nativeLink` command?

**Comment by reid-spencer** [comment]
I concur that this is a duplicate of #2024. I've installed @JD557 's workaround with success.  A real fix should still be implemented so such workarounds are unnecessary.
